tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc

```diff
@@ -0,0 +1,437 @@
+  if (!client_->IsConnected()) {
+    Status status = client_->Connect();
+    if (!status.ok()) return status;
+
+    status = Handshake();
+    if (!status.ok()) {
+      Status disconnect_status = client_->Disconnect();
+      if (!disconnect_status.ok()) LOG(ERROR) << disconnect_status.ToString();
+
+      return status;
+    }
+  }
+
+  return Status::OK();
+}
+
+Status IgniteDatasetIterator::CloseConnection() {
+  if (cursor_id_ != -1 && !last_page_) {
+    Status conn_status = EstablishConnection();
+    if (!conn_status.ok()) return conn_status;
+
+    TF_RETURN_IF_ERROR(client_->WriteInt(18));  // Message length
+    TF_RETURN_IF_ERROR(
+        client_->WriteShort(close_connection_opcode));   // Operation code
+    TF_RETURN_IF_ERROR(client_->WriteLong(0));           // Request ID
+    TF_RETURN_IF_ERROR(client_->WriteLong(cursor_id_));  // Resource ID
+
+    int32_t res_len;
+    TF_RETURN_IF_ERROR(client_->ReadInt(&res_len));
+    if (res_len < 12)
+      return errors::Internal("Close Resource Response is corrupted");
+
+    int64_t req_id;
+    TF_RETURN_IF_ERROR(client_->ReadLong(&req_id));
+    int32_t status;
+    TF_RETURN_IF_ERROR(client_->ReadInt(&status));
+    if (status != 0) {
+      uint8_t err_msg_header;
+      TF_RETURN_IF_ERROR(client_->ReadByte(&err_msg_header));
+      if (err_msg_header == string_val) {
+        int32_t err_msg_length;
+        TF_RETURN_IF_ERROR(client_->ReadInt(&err_msg_length));
+        uint8_t* err_msg_c = new uint8_t[err_msg_length];
+        TF_RETURN_IF_ERROR(client_->ReadData(err_msg_c, err_msg_length));
+        std::string err_msg((char*)err_msg_c, err_msg_length);
+        delete[] err_msg_c;
+
+        return errors::Internal("Close Resource Error [status=", status,
+                                ", message=", err_msg, "]");
+      }
+      return errors::Internal("Close Resource Error [status=", status, "]");
+    }
+
+    LOG(INFO) << "Query Cursor " << cursor_id_ << " is closed";
+
+    cursor_id_ = -1;
+
+    return client_->Disconnect();
+  } else {
+    LOG(INFO) << "Query Cursor " << cursor_id_ << " is already closed";
+  }
+
+  return client_->IsConnected() ? client_->Disconnect() : Status::OK();
+}
+
+Status IgniteDatasetIterator::GetNextInternal(IteratorContext* ctx,
+                                              std::vector<Tensor>* out_tensors,
+                                              bool* end_of_sequence) {
+  if (remainder_ == 0 && last_page_) {
+    LOG(INFO) << "Query Cursor " << cursor_id_ << " is closed";
+
+    cursor_id_ = -1;
+    *end_of_sequence = true;
+    return Status::OK();
+  } else {
+    Status status = EstablishConnection();
+    if (!status.ok()) return status;
+
+    if (remainder_ == -1 || remainder_ == 0) {
+      Status status = remainder_ == -1 ? ScanQuery() : LoadNextPage();
+      if (!status.ok()) return status;
+    }
+
+    uint8_t* initial_ptr = ptr_;
+    std::vector<int32_t> types;
```

Comment:
By the above comment, I think this vector is unused. However, since the data could have types that don't match the schema, we need some validation here, to check that `tensors` match the output types and shapes implied by the schema.

Reply:
Thanks, as I wrote in comment above I forgot about it. I've added `types` filling and correspondent check. Please, see `CheckTypes` method.